all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,505 @@
+# External Client Configuration
+
+Notes for this proposal:
+
+* Everything is subject to change both during proposal time and afterwards during implementation.
+* For each place a decision is made, a "💭 Why X? ..." aside is present explaining why.
+* For each place an open question exists, a "❓ X?" exists to ask it.
+
+## Overview
+
+There are many ways to access Temporal programmatically including CLI, SDK, and other components that leverage those
+such as samples and Terraform providers.
+
+Temporal needs a way for users to provide client configuration without hardcoding in CLI arguments or SDK options. This
+will allow code to seamlessly switch between profiles such as between a dev-server and cloud. The same configuration
+must work across all SDKs/CLIs.
+
+### Why?
+
+What are the reasons for building this?
+
+* Users want to switch environments without changing their code, but today they cannot.
+* Our samples should be able to be run on multiple environments including cloud, but today they are localhost only.
+
+### Goals
+
+* Common, well-specified/type-safe configuration format usable by Temporal tooling for client options and usable
+  manually by advanced users
+* Implementation in all Temporal SDKs for reading and writing of this configuration
+* All Temporal CLI/SDKs updated to have easy client creation using this configuration
+* Common on-disk location in every platform for reading/writing this format
+* Ability to provide environment variables to represent values in the configuration
+* Well-specified hierarchy for on-disk overridable by environment variables overridable by in-CLI/SDK values
+* 0 new dependencies in SDKs
+  * 💭 Why?
+    * Dependencies in our SDKs subject our users to these dependencies and their version constraints, so we should avoid
+      this as much as possible.
+  * ❓ Can we just have an "extension" that has new dependencies?
+    * Yes, but ideally users don't have to add dependencies on extension libraries just to get this functionality.
+* Profile names within the configuration
+  * 💭 Why?
+    * It is common in configurations like this to want to be able to share a single file with many configurations.
+
+### Future Goals
+
+None identified at this time. The goals are simple enough for the MVP and the non-goals are ones we never want.
+
+### Non-goals
+
+* Common format for declarative workers, schedules, etc. This is only for client options.
```

Comment:
Let's go slowly here. In the future, if we want to support declarative configuration of other aspects of Temporal, then we will not want multiple config files. The natural design decision at this point is to say

> We're currently designing for client configuration only. But the design we've chosen allows the possibility that other top-level keys will be added in the future.

In which case client config should be under its own top-level key, allowing other top-level keys in the future for all the things we're deliberately not designing or thinking about today: settings, schedules, etc

Reply:
I think declarative resources should be tackled separately. Having said that, we can probably add non-client things in the future. But I'm not sure we need to qualify client things. The reason this is CLI/SDK configuration and not client configuration is because you need these things to basically do anything with the CLI/SDK. From a user POV it's just "config". If we want to add things that do not apply to clients (e.g. worker configuration) that may make sense, but I still think these general options can be top level.

I think asking all users to add this `client.` prefix for our own personal future proofing is a bit rough. IMO it is a worse developer experience to make users use `TEMPORAL_CLIENT_ADDRESS` and `client.address` and what not over `TEMPORAL_ADDRESS` and `address`. These latter have been in use for years already. I don't thin we should harm developer experience here over a perceived future.

Reply:
> I think declarative resources should be tackled separately. Having said that, we can probably add non-client things in the future.

The proposed file name is `<app-config-dir>/temporalio/temporal.json`, so this file clearly must include all temporal config. And that's a good thing! We don't want multiple files.

> But I'm not sure we need to qualify client things.

In the proposed schema, all the keys related to client config are already nested below `profiles`, and `name`. There are currently 6 of those keys: `address`, `namespace`, `api_key`, `tls`, `codec`, `grpcMeta`.

But 6 is just the number in the first iteration of the proposal; it's bound to grow over time, and the result will be that the other sections we add to the file will be their own root nodes, swimming oddly among a sea of client config details. If we look at file formats such as TOML and TOML-ish things like `~/.gitconfig`, there are no top-level keys.

I can only see advantage and no disadvantage in grouping the client config under its own key (e.g. `clientConfig`), especially since we're already two levels in. Can you list the disadvantages you see with that?

Reply:
> will be that the other sections we add to the file will be their own root nodes, swimming oddly among a sea of client config details. 

But there are no other sections today. I don't agree we should harm developer experience today for some may-never-happen unrealized future.

> Can you list the disadvantages you see with that?

I mentioned this in the last paragraph above:

> I think asking all users to add this `client.` prefix for our own personal future proofing is a bit rough. IMO it is a worse developer experience to make users use `TEMPORAL_CLIENT_ADDRESS` and `client.address` and what not over `TEMPORAL_ADDRESS` and `address`. These latter have been in use for years already. I don't think we should harm developer experience here over a perceived future that may not come about.

I don't want everyone to qualify these things as client-specific config just because there is technically the possibility of some non-client future.